[MODULAR] Status Indicators with Preference

nevimer opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

This is a port from Citadel RP / Polaris / Vorestation. This pull request features new sprites from [AtoriBirb](https://github.com/AtoriBirb)!

## How This Contributes To The Skyrat Roleplay Experience

Let's face it: Combat is confusing. A current Skyrat edit we have is adding an inverse drop shadow on people who are stam crit, but it's extremely hard to tell things about people at a glance, and combat is so fast... and confusion occurs as to WHAT is going on, often. Combat indicators added by this PR aim to add visual clarity to situations, make you understand someone's status in visually unclear situations. For players with limited screen real estate, or poor eyesight, this is especially useful. This is FAR better than shift clicking everyone in combat. 


Nothing about this PR adds any information you can't gain through shift click inspecting a mob, and all indicators aim to have the same functionality and behaviour as runechat - a toggleable preference if you prefer to keep this information to the chat bar (just like you can play without runechat (how)). 

## Media


https://user-images.githubusercontent.com/77420409/163046612-14ab98b5-07c7-41de-958f-412737e37087.mp4


https://user-images.githubusercontent.com/77420409/163046509-a086c817-a400-4b66-b7cc-01df6ba1b461.mp4






## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: [nevimer](https://github.com/nevimer), [AtoriBirb](https://github.com/AtoriBirb)
add: Status Indicators! Toggle in preferences!
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

softcerv commented:
This is definitely a great change, especially for newer players!

Gandalf2k15 commented on code/datums/status_effects/debuffs.dm:
```diff
@@ -160,12 +168,16 @@
 		tick_interval = -1
 	RegisterSignal(owner, SIGNAL_ADDTRAIT(TRAIT_SLEEPIMMUNE), .proc/on_owner_insomniac)
 	RegisterSignal(owner, SIGNAL_REMOVETRAIT(TRAIT_SLEEPIMMUNE), .proc/on_owner_sleepy)
```
Move these all to a modular file and use overrides

Gandalf2k15 commented on code/datums/status_effects/debuffs.dm:
```diff
@@ -33,11 +33,12 @@
 	ADD_TRAIT(owner, TRAIT_IMMOBILIZED, TRAIT_STATUS_EFFECT(id))
 	ADD_TRAIT(owner, TRAIT_HANDS_BLOCKED, TRAIT_STATUS_EFFECT(id))
-
+	owner.add_status_indicator("stunned") // SKYRAT EDIT ADDITION: Status indicators
```
As in these instances

nevimer replied:
I don't understand what you mean by use overrides

Gandalf2k15 replied:
See what I sent you in discord

Gandalf2k15 commented on code/datums/status_effects/debuffs.dm:
```diff
@@ -241,6 +255,7 @@
 	ADD_TRAIT(owner, TRAIT_IMMOBILIZED, TRAIT_STATUS_EFFECT(id))
 	ADD_TRAIT(owner, TRAIT_HANDS_BLOCKED, TRAIT_STATUS_EFFECT(id))
 	ADD_TRAIT(owner, TRAIT_NUMBED, "stasis") //SKYRAT EDIT START - STASIS APPLIES NUMBING
+	owner.add_status_indicator("paralysis") // SKYRAT EDIT ADDITION: Status indicators
```
Make all of these defines. For instance STATUS_INDICATOR_PARALYSIS

GoldenAlpharex commented:
Personally not a big fan of this, mostly for the reason that those things are something that you wouldn't see at a glance, and something that you have to trade a bit of your attention to figure out.

In my opinion, these are *too* big and obvious. There's no subtlety to them, it's something OOC that's appearing IC. Like, we currently only have one (1) OOC "status" element and it's the combat indicator, and the only reason why people know not to interpret it as something IC is because it's included in the application process.

I don't mind the addition of visual indicators that could be taken in-character, but I just know that this is 100% something I'd immediately turn off because it's just not great for immersion to have giant circles floating above people's head, and too distracting from what's really important, the people underneath.

I already mentioned that on the Discord a few weeks ago, I'm just relaying it here so it doesn't get lost to time.

nevimer commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -356,7 +356,7 @@
 	visible_message(span_danger("[src] manages to [cuff_break ? "break" : "remove"] [I]!"))
 	to_chat(src, span_notice("You successfully [cuff_break ? "break" : "remove"] [I]."))
-	
+
```
I can't seem to undo this. Weh.

nevimer commented:
> Personally not a big fan of this, mostly for the reason that those things are something that you wouldn't see at a glance, and something that you have to trade a bit of your attention to figure out.
> 
> In my opinion, these are _too_ big and obvious. There's no subtlety to them, it's something OOC that's appearing IC. Like, we currently only have one (1) OOC "status" element and it's the combat indicator, and the only reason why people know not to interpret it as something IC is because it's included in the application process.
> 
> I don't mind the addition of visual indicators that could be taken in-character, but I just know that this is 100% something I'd immediately turn off because it's just not great for immersion to have giant circles floating above people's head, and too distracting from what's really important, the people underneath.
> 
> I already mentioned that on the Discord a few weeks ago, I'm just relaying it here so it doesn't get lost to time.

Your concerns about it are valid, and I appreciate the feedback on it. I ultimately saw it fit to port due to the nature of it's origins - servers using 15x15 viewports and HRP environments. I liken it's ICK-OCK rating to be that equal to runechat - it's something there for the viewer, not your character - an aid to keep your eyes in the game and off the chat box for spammy things - leaving the chat box for comms/warnings/logs, instead of spilling it will character inspections (where I start to miss comms in combat). 

With the changes at TG strafing towards not killing people, and manipulating them in several statuses (such as brainwashing, sacrificing), and the rapid speed of TG combat and OOC confusion potential, I hope to have the gap between OOC misreading the situation and IC misreading the situation blurred, and help players make the acts they intended to make, not the one they they actually made at the time.

For example, security officers overkilling suspects while they're stamcrit, downed, or otherwise. Additionally, it could help bring clarity to other situations where status effects that are poorly represented are used. 

I hope this addresses your concerns slightly!

nevimer commented:
> > Personally not a big fan of this, mostly for the reason that those things are something that you wouldn't see at a glance, and something that you have to trade a bit of your attention to figure out.
> > In my opinion, these are _too_ big and obvious. There's no subtlety to them, it's something OOC that's appearing IC. Like, we currently only have one (1) OOC "status" element and it's the combat indicator, and the only reason why people know not to interpret it as something IC is because it's included in the application process.
> > I don't mind the addition of visual indicators that could be taken in-character, but I just know that this is 100% something I'd immediately turn off because it's just not great for immersion to have giant circles floating above people's head, and too distracting from what's really important, the people underneath.
> > I already mentioned that on the Discord a few weeks ago, I'm just relaying it here so it doesn't get lost to time.
> 
> Your concerns about it are valid, and I appreciate the feedback on it. I ultimately saw it fit to port due to the nature of it's origins - servers using 15x15 viewports and HRP environments. I liken it's ICK-OCK rating to be that equal to runechat - it's something there for the viewer, not your character - an aid to keep your eyes in the game and off the chat box for spammy things - leaving the chat box for comms/warnings/logs, instead of spilling it will character inspections (where I start to miss comms in combat).
> 
> With the changes at TG strafing towards not killing people, and manipulating them in several statuses (such as brainwashing, sacrificing), and the rapid speed of TG combat and OOC confusion potential, I hope to have the gap between OOC misreading the situation and IC misreading the situation blurred, and help players make the acts they intended to make, not the one they they actually made at the time.
> 
> For example, security officers overkilling suspects while they're stamcrit, downed, or otherwise. Additionally, it could help bring clarity to other situations where status effects that are poorly represented are used.
> 
> I hope this addresses your concerns slightly!

Oh, also, funny story. When I ported runechat to a certain HRP place, players treated runechat IC lol. They were all "WHAT ARE THESE WORDS OVER MY HEAD!!" on the radio (it was hilarious at the time).

tf-4 commented:
Not a fan of being able to see exactly when someone is about to start getting up from a knockdown. Takes some nuance and knowledge out of combat.

nevimer commented:
> Not a fan of being able to see exactly when someone is about to start getting up from a knockdown. Takes some nuance and knowledge out of combat.

They actually lag behind in the current commit.  By the time the indicator is removed, the person will be standing up already!

tf-4 commented:
> > Not a fan of being able to see exactly when someone is about to start getting up from a knockdown. Takes some nuance and knowledge out of combat.
> 
> They actually lag behind in the current commit. By the time the indicator is removed, the person will be standing up already!

Complaint revoked, then. Still in half a mind to agree with Golden, but if you can toggle it off entirely then I don't really see the harm.

GoldenAlpharex commented:
I think a first good step at making them less imposing would be to give a try at not having them be in bubbles, but rather be just an icon with no background, which would lessen their footprint on the screen quite a lot.

nevimer commented:
> 

now featuring new sprites!

AtoriBirb commented:
Gross looks like a bird sprited them

theOOZ commented:
I have these disabled with the pref but I still see them.

nevimer commented:
> I have these disabled with the pref but I still see them.

Will take a look for you! Any other combo of things enabled / disabled?

theOOZ commented:
Nothing special besides may be related to silicons.

GoldenAlpharex commented:
I'm drafting this because there's still changes that need to be made to it, and since it's a lot of content, I'd also like to review it to ensure the code is up to standards, because I know there's a lot of calls on `life()` in there and it's not something that makes me overly happy.

nevimer commented:
@GoldenAlpharex Sure!

In my next commit which I will push after a bit of local testing, I switched it to not run on life ticks, instead, status effect applications and COMSIG_CARBON_HEALTH_UPDATE signals. Should cut down on how often it is ran for mobs not in combat as well as _HOPEFULLY_ resolve the ghost images bug.

nevimer commented:
https://user-images.githubusercontent.com/77420409/165263966-99e4d332-a0b6-44ce-a837-2ed2b0d662fd.mp4

Alright, we use signals now and on_apply/remove

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/overrides.dm:
```diff
@@ -0,0 +1,78 @@
+	owner.remove_status_indicator("sleeping") // SKYRAT EDIT ADDITION: Status indicators
+
+
+
+
+
+/*
+/datum/status_effect/incapacitating/sleeping/on_apply()
+	. = ..()
+	if(!HAS_TRAIT(owner, TRAIT_SLEEPIMMUNE))
+		owner.add_status_indicator("sleeping") // SKYRAT EDIT ADDITION: Status indicators
+/datum/status_effect/incapacitating/sleeping/on_remove()
+	. = ..()
+	if(!HAS_TRAIT(owner, TRAIT_SLEEPIMMUNE))
+		owner.remove_status_indicator("sleeping") // SKYRAT EDIT ADDITION: Status indicators
+*/
+
+
+/datum/status_effect/grouped/stasis/on_apply()
+	. = ..()
+	owner.add_status_indicator("paralysis") // SKYRAT EDIT ADDITION: Status indicators
+/datum/status_effect/grouped/stasis/on_remove()
+	. = ..()
+	owner.remove_status_indicator("paralysis") // SKYRAT EDIT ADDITION: Status indicators
```
Again, here, is it possible that the `paralysis` can cause issues with the other source of paralysis, and cause it to either be removed before it should be removed, or to stay behind when it shouldn't be?

I'd probably suggest trying to go for both a status indicator name, as well as an indicator source, to avoid race conditions.
```suggestion
/datum/status_effect/grouped/stasis/on_apply()
	. = ..()
	owner.add_status_indicator("paralysis")

/datum/status_effect/grouped/stasis/on_remove()
	. = ..()
	owner.remove_status_indicator("paralysis")
```

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+#define PLANE_STATUS -12 //Status Indicators that show over mobs' heads when certain things like stuns affect them.
+
+#define VIS_STATUS			24
+
+#define VIS_COUNT			24 //Must be highest number from above.
```
What do these stand for? Some documentation is really handy in stuff involving rendering.
```suggestion
#define VIS_STATUS 24
#define VIS_COUNT 24 //Must be highest number from above.
```

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+		status.alpha = 255
+	else
+		status.alpha = 0
+/mob
+	var/status_enabled = TRUE
+/mob/living
+	var/list/status_indicators = null // Will become a list as needed.
```
I'd also like for you to document these variables, since you're adding them. For instance, I'm not sure what kind of list `status_indicator` is supposed to be, if it's supposed to be a list of strings or a list of images.

Also, what even is the point of `status_enabled` here? I haven't seen it come up anywhere else in your code so far.
```suggestion
/mob
	var/status_enabled = TRUE

/mob/living
	var/list/status_indicators = null // Will become a list as needed.
```

nevimer replied:
status_enabled gone next commit

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+/mob/living
+	var/list/status_indicators = null // Will become a list as needed.
+
+/* /mob/living/carbon/proc/is_critical()
+	if(HAS_TRAIT(src, TRAIT_CRITICAL_CONDITION))
+		return TRUE */
+
```
You forgot to remove some commented code.
```suggestion
```

nevimer replied:
gone in next commit

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	if(HAS_TRAIT(src, TRAIT_CRITICAL_CONDITION))
+		return TRUE */
+
+/mob/living/proc/is_weakened()
```
Some documentation before your proc, please.
With `/// `, because that's how AUTODOC works.
So basically, something like
```
/// Returns TRUE if the mob is weakened, null if not.
```

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	if(HAS_TRAIT_FROM(src, TRAIT_FLOORED, TRAIT_STATUS_EFFECT(STAT_TRAIT)) || has_status_effect(/datum/status_effect/incapacitating/knockdown))
+		return TRUE
+
+/mob/living/proc/is_stunned()
```
Ditto, documentation please.

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	if(HAS_TRAIT_FROM(src, TRAIT_INCAPACITATED, TRAIT_STATUS_EFFECT(STAT_TRAIT)) || HAS_TRAIT_FROM(src, TRAIT_IMMOBILIZED, TRAIT_STATUS_EFFECT(STAT_TRAIT)) || HAS_TRAIT_FROM(src, TRAIT_IMMOBILIZED, CHOKEHOLD_TRAIT))
+		return TRUE
+
+/mob/living/proc/is_paralyzed()
```
Ditto, documentation please.

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	if(HAS_TRAIT_FROM(src, TRAIT_FLOORED, CHOKEHOLD_TRAIT) || HAS_TRAIT(src, TRAIT_CRITICAL_CONDITION) || HAS_TRAIT_FROM(src, TRAIT_INCAPACITATED, STAMINA) || HAS_TRAIT_FROM(src, TRAIT_INCAPACITATED, TRAIT_STATUS_EFFECT(STAT_TRAIT)) && HAS_TRAIT_FROM(src, TRAIT_IMMOBILIZED, TRAIT_STATUS_EFFECT(STAT_TRAIT)) && HAS_TRAIT_FROM(src, TRAIT_FLOORED, TRAIT_STATUS_EFFECT(STAT_TRAIT)))
+		return TRUE
+
+/mob/living/proc/is_unconcious()
```
Ditto, documentation please.

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	if(HAS_TRAIT(src, TRAIT_KNOCKEDOUT))
+		return TRUE
+
+/mob/living/proc/is_confused()
```
Ditto, documentation please.

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+/mob/living/proc/is_confused()
+	if(has_status_effect(/datum/status_effect/confusion))
+		return TRUE
+
+/* /mob/living/carbon/proc/is_blind_status()
+	if(HAS_TRAIT_FROM(src, TRAIT_BLIND, STATUS_EFFECT_TRAIT) || HAS_TRAIT_FROM(src, TRAIT_BLIND, EYES_COVERED) && !HAS_TRAIT_FROM(src, TRAIT_BLIND, QUIRK_TRAIT)) // We don't want permanently blind users (those who took the trait) to have this forever.
+		return TRUE */ // Disabled, uncomment if we want this.
```
```suggestion
```

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	if(HAS_TRAIT_FROM(src, TRAIT_BLIND, STATUS_EFFECT_TRAIT) || HAS_TRAIT_FROM(src, TRAIT_BLIND, EYES_COVERED) && !HAS_TRAIT_FROM(src, TRAIT_BLIND, QUIRK_TRAIT)) // We don't want permanently blind users (those who took the trait) to have this forever.
+		return TRUE */ // Disabled, uncomment if we want this.
+
+/mob/living/carbon/death(gibbed) // On death, we clear the indiciators
+	..() // Call the TG death. Do not . = ..()!
+	for(var/iteration in status_indicators) // When we die, clear the indicators.
+		remove_status_indicator(icon_state) // The indicators are named after their icon_state and type
```
Why are you removing them only on `/mob/living/carbon` when you're adding them to `/mob/living`?

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	..() // Call the TG death. Do not . = ..()!
+	for(var/iteration in status_indicators) // When we die, clear the indicators.
+		remove_status_indicator(icon_state) // The indicators are named after their icon_state and type
+/* /datum/status_effect/incapacitating/on_apply()
+	. = ..()
+	addtimer(CALLBACK(owner, /mob/living/.proc/status_sanity), 100 MILLISECONDS)
+/datum/status_effect/incapacitating/on_remove()
+	. = ..()
+	addtimer(CALLBACK(owner, /mob/living/.proc/status_sanity), 1000 MILLISECONDS) */
+
+/* /mob/living/carbon/apply_effect()
+	. = ..()
+	status_sanity() */
```
```suggestion
```

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	. = ..()
+	RegisterSignal(src, COMSIG_CARBON_HEALTH_UPDATE, .proc/status_sanity)
+
+/* /mob/living/updatehealth()
+	. = ..()
+	status_sanity() */
```
```suggestion
```

nevimer replied:
gone next commit

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+
+/mob/living/carbon/Initialize(mapload)
+	. = ..()
+	RegisterSignal(src, COMSIG_CARBON_HEALTH_UPDATE, .proc/status_sanity)
```
Remember to UnregisterSignal() on Delete, to avoid hard-deletions.

nevimer replied:
don't know how to do that

nevimer replied:
I will have to be shown a good spot to add this onto, but not death as mobs do revive.

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	if(stat == DEAD)
+		for(var/iteration in status_indicators) // When we die, clear the indicators.
+			remove_status_indicator(icon_state) // The indicators are named after their icon_state and type
+	is_weakened() ? add_status_indicator("weakened") : remove_status_indicator("weakened")
+	is_paralyzed() ? add_status_indicator("paralysis") : remove_status_indicator("paralysis")
+	is_stunned() ? add_status_indicator("stunned") : remove_status_indicator("stunned")
+	is_unconcious() ? add_status_indicator("sleeping") : remove_status_indicator("sleeping")
+	is_confused() ? add_status_indicator("confused") : remove_status_indicator("confused")
```
Even if this works, this is bad for readability and doesn't really follow the coding standards.

Each of these should be their own proc, called `handle_status_indicator_status_name()` (so like `handle_status_weakened()`) and the code inside of these procs should be something like this:
```
/// A simple proc to handle either adding or removing the weakened status indicator.
/mob/living/proc/handle_status_indicator_weakened()
	if(is_weakened())
		add_status_indicator(INDICATOR_STATUS_WEAKENED)
		return
	remove_status_indicator(INDICATOR_STATUS_WEAKENED)
```

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+/* /mob/living/updatehealth()
+	. = ..()
+	status_sanity() */
+/mob/living/proc/status_sanity()
```
Document this proc too, please.

nevimer replied:
done next commit

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	status_sanity() */
+/mob/living/proc/status_sanity()
+	SIGNAL_HANDLER
+	if(stat == DEAD)
+		for(var/iteration in status_indicators) // When we die, clear the indicators.
+			remove_status_indicator(icon_state) // The indicators are named after their icon_state and type
```
This also probably could be its own proc, like, the for loop to remove them all, it should be something like `remove_all_status_indicators()`.

You also need to add a `return` there, otherwise they will, in fact, be re-applied regardless afterwards.
```suggestion
	if(stat == DEAD)
		for(var/iteration in status_indicators) // When we die, clear the indicators.
			remove_status_indicator(icon_state) // The indicators are named after their icon_state and type
		return
```

nevimer replied:
well true, yeah. Death I am sure removes all status effects anyways but done next commit

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	is_stunned() ? add_status_indicator("stunned") : remove_status_indicator("stunned")
+	is_unconcious() ? add_status_indicator("sleeping") : remove_status_indicator("sleeping")
+	is_confused() ? add_status_indicator("confused") : remove_status_indicator("confused")
+//	is_blind_status() ? add_status_indicator("blinded") : remove_status_indicator("blinded") // Disabled, keeping for IF we decide we want this.
+//	is_critical() ? add_status_indicator("critical") : remove_status_indicator("critical") // Also disabled. Crosses medhud territory, despite being a shift click examine thing.
```
No need to keep this in, it's really easy to add back in case it's wanted in the future.
```suggestion
```

nevimer replied:
gone next commit

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+//	is_blind_status() ? add_status_indicator("blinded") : remove_status_indicator("blinded") // Disabled, keeping for IF we decide we want this.
+//	is_critical() ? add_status_indicator("critical") : remove_status_indicator("critical") // Also disabled. Crosses medhud territory, despite being a shift click examine thing.
+
+/mob/living/proc/add_status_indicator(image/thing)
```
Document this, please.

Also, as I stated earlier, I think this should have two arguments, the second being an `indicator_source`, so that when you add an indicator, you can specify its source and not duplicate it if you already have one, and instead just add to the list of its sources, and then you can delete entries from the list of sources of said indicator if it exists, and then just remove the status indicator if its list of sources reaches a length of zero, like we can do for traits.

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+/mob/living/proc/add_status_indicator(image/thing)
+	if(get_status_indicator(thing)) // No duplicates, please.
+		return
+	if(!istype(thing, /image))
+		thing = image(icon = 'modular_skyrat/master_files/icons/mob/status_indicators.dmi', loc = src, icon_state = thing)
```
From what I could see so far, the parameter `thing` (which should ***definitely*** have a better name, like `indicator_name`) is never an image, and always seems to be a string. Thus, this if() statement feels pointless, because it's always going to be the icon_state that you pass to it.

I feel like those images could be cached, and then added as overlays on the mob afterwards, so basically you'd likely go from having up to like five images for every mob, down to five images in total, if it truly works like I'm feeling it does. I'll ask around to see if that's how it works, but you can still try that in the meantime, if that's faster for you.

GoldenAlpharex replied:
I asked around on /tg/, and apparently, yes, it's possible to add the same image as an overlay to many different mobs. So this is what I'd like you to do, that way there's not an ever-increasing memory cost tied to these indicators, based on the amount of mobs that exist, but rather a flat cost of up to 5 images for the round. I think we can both agree that's a very good thing for performance.

nevimer replied:
renamed next commit, not sure how to do that

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	handle_status_indicators()
+
+// Similar to above but removes it instead, and nulls the list if it becomes empty as a result.
+/mob/living/proc/remove_status_indicator(image/thing)
+	thing = get_status_indicator(thing)
+
+	cut_overlay(thing)
+	LAZYREMOVE(status_indicators, thing)
+	handle_status_indicators()
```
Document this properly, and rename `thing` to be something less vague, because that's an awful name for it.

And, in general, I really don't think that argument should be an `/image`, it should just be a string.

nevimer replied:
renamed next commit

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	LAZYREMOVE(status_indicators, thing)
+	handle_status_indicators()
+
+/mob/living/proc/get_status_indicator(image/thing)
+	if(!istype(thing, /image))
+		for(var/image/I in status_indicators)
+			if(I.icon_state == thing)
+				return I
+	return LAZYACCESS(status_indicators, LAZYFIND(status_indicators, thing))
```
Ditto, document this and give `thing` a better name, while also making it not a `/image`. You could also even go a step further and add `RETURN_TYPE(/image)`, which I'll show you how to do, and that allows you to access variables on the return value of that proc without having to store it in a variable that's typed to `/image`.

Here's a quick suggestion to clear that single-letter variable because that's also pretty bad. (We use `as anything` because we know there's only going to be `/image`-type elements in that list (if not, that's something you should fix).

Mind you, I didn't remove the fact that they can be images, mostly because I'm not certain of how it could break your code.
```suggestion
/mob/living/proc/get_status_indicator(image/thing)
	RETURN_TYPE(/image)
	if(!istype(thing, /image))
		for(var/image/indicator as anything in status_indicators)
			if(indicator.icon_state == thing)
				return indicator
	return LAZYACCESS(status_indicators, LAZYFIND(status_indicators, thing))
```

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+				return I
+	return LAZYACCESS(status_indicators, LAZYFIND(status_indicators, thing))
+
+/mob/living/proc/cut_indicators_overlays()
```
Document this proc, please!

nevimer replied:
documented next commit

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+
+/mob/living/proc/cut_indicators_overlays()
+	if(!status_indicators) // sometimes the overlay cutting misses, so if theres nothing when its called, lets just clear them all!
+		var/list/static/potential_indicator = list(
+			"weakened",
+			"paralysis",
+			"stunnded",
+			"sleeping",
+			"confused",
+			"blinded",
+			"critical")
```
You were very close from formatting it properly, but here's how it's supposed to look like.

Furthermore, you still had blinded and critical in there, which aren't a thing anymore.

Lastly, as I mentioned earlier, all of these should be defines, so that if we change the value inside of it later, it gets changed everywhere at once.
```suggestion
		var/list/static/potential_indicator = list(
			"weakened",
			"paralysis",
			"stunnded",
			"sleeping",
			"confused",
		)
```

nevimer replied:
removed next commit cause i thought of something better

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+			"confused",
+			"blinded",
+			"critical")
+		for(potential_indicator in overlays)
+			cut_overlay(potential_indicator)
```
I feel like this isn't going to work, considering how you're trying to find a string in a list of images.

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+			cut_overlay(potential_indicator)
+
+	else
+		for(var/thing in status_indicators)
+			cut_overlay(thing)
```
Mind you, I might not be sure of how this works, but I really hope you're trying to remove images from the overlay and not just strings, because I feel like that could be an issue too.

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+
+	if(stat == DEAD)
+		return
+	var/mob/living/carbon/carbon = src
+	var/icon_scale = get_icon_scale(carbon)
```
Wouldn't `get_icon_scale(src)` also work, or is it a carbon-exclusive thing? If it's the latter, then I feel like all of this code should be moved to the carbon level, but I somehow feel like it wouldn't be a carbon-exclusive proc.

nevimer replied:
Correct, icon scale is a carbon var, but any mob carries statuses.

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+
+	// In /mob/living's `update_transform()`, the sprite is horizontally shifted when scaled up, so that the center of the sprite doesn't move to the right.
+	// Because of that, this adjustment needs to happen with the future indicator row as well, or it will look bad.
+	current_x_position -= (32 / 2) * (icon_scale - 1)
```
Where does that `32 / 2` come from? Should likely be a define, because that's not going to change between the times the proc is ran.

nevimer replied:
Gone next commit!

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+		for(var/thing in status_indicators)
+			cut_overlay(thing)
+
+// Refreshes the indicators over a mob's head. Should only be called when adding or removing a status indicator with the above procs,
+// or when the mob changes size visually for some reason.
+/mob/living/proc/handle_status_indicators()
```
This does look like documentation, but in order for it to be registered as documentation, please use `/// ` and not `// `

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	current_x_position -= (32 / 2) * (icon_scale - 1)
+
+	// Now the indicator row can actually be built.
+	for(var/thing in status_indicators)
+		var/image/I = thing
+
+		// This is a semi-HUD element, in a similar manner as medHUDs, in that they're 'above' everything else in the world,
+		// but don't pierce obfuscation layers such as blindness or darkness, unlike actual HUD elements like inventory slots.
+		I.plane = PLANE_STATUS
+		I.layer = STATUS_LAYER
+		I.appearance_flags = PIXEL_SCALE|TILE_BOUND|NO_CLIENT_COLOR|RESET_COLOR|RESET_ALPHA|RESET_TRANSFORM|KEEP_APART
+		I.pixel_y = y_offset
+		I.pixel_x = current_x_position
+		add_overlay(I)
+		// Adding the margin space every time saves a conditional check on the last iteration,
+		// and it won't cause any issues since no more icons will be added, and the var is not used for anything else.
+		current_x_position += STATUS_INDICATOR_ICON_X_SIZE + STATUS_INDICATOR_ICON_MARGIN
```
Your list of `status_indicator` should already be a list of `/image`, so you just typecast it in the for-loop, and add an `as anything` so that DM doesn't do an `istype()` check for each element in `status_indicator`, since we know that they're all images. Performance improvements!

Also, single-letter variables are bad, don't do that.

Also, the `appearance_flags` here should probably be in a define too, so that it's easy to change in the future, and that it makes the whole thing more readable.

Also, if you're just adding two defines together, just make a third define that's already adding the two of them together, so you end up just adding one thing on line 180, which makes it more readable and coherent.
```suggestion
	for(var/image/indicator as anything in status_indicators)
		// This is a semi-HUD element, in a similar manner as medHUDs, in that they're 'above' everything else in the world,
		// but don't pierce obfuscation layers such as blindness or darkness, unlike actual HUD elements like inventory slots.
		indicator.plane = PLANE_STATUS
		indicator.layer = STATUS_LAYER
		indicator.appearance_flags = PIXEL_SCALE | TILE_BOUND | NO_CLIENT_COLOR | RESET_COLOR | RESET_ALPHA | RESET_TRANSFORM | KEEP_APART
		indicator.pixel_y = y_offset
		indicator.pixel_x = current_x_position
		add_overlay(indicator)
		// Adding the margin space every time saves a conditional check on the last iteration,
		// and it won't cause any issues since no more icons will be added, and the var is not used for anything else.
		current_x_position += STATUS_INDICATOR_ICON_X_SIZE + STATUS_INDICATOR_ICON_MARGIN
```

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+		// and it won't cause any issues since no more icons will be added, and the var is not used for anything else.
+		current_x_position += STATUS_INDICATOR_ICON_X_SIZE + STATUS_INDICATOR_ICON_MARGIN
+
+/mob/living/proc/get_icon_scale(carbon)
```
Please document this proc.

Also, you really shouldn't be doing what you're doing here, you should pass no argument and make this proc just return 1, and override this proc on `mob/living/carbon` that would look like this:
```
/mob/living/carbon/get_icon_scale()
	return dna.current_body_size ? passed_mob.dna.current_body_size : 1
```

This is makes more sense in terms of inheritance, and as such will also be more performant, too!

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+	var/mysize = (passed_mob.dna.current_body_size ? passed_mob.dna.current_body_size : 1)
+	return mysize
+
+/atom/movable/screen/plane_master/runechat/status
+	name = "status plane master"
```
Is there any reason why this is a subtype of `/plane_master/runechat`?

Also, rename it to be clearer:
```suggestion
/atom/movable/screen/plane_master/runechat/status
	name = "status indicator plane master"
```

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+
+/atom/movable/screen/plane_master/runechat/status
+	name = "status plane master"
+	plane = PLANE_STATUS
```
This should be renamed to this:
```suggestion
	plane = PLANE_STATUS_INDICATOR
```

GoldenAlpharex commented on tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/skyrat/status_indicators.tsx:
```diff
@@ -0,0 +1,8 @@
+import { CheckboxInput, FeatureToggle } from "../../base";
+
+export const enable_status_indicators: FeatureToggle = {
+  name: "Show/Hide Status Indicators",
```
```suggestion
  name: "Display Status Indicators",
```

GoldenAlpharex commented on tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/skyrat/status_indicators.tsx:
```diff
@@ -0,0 +1,8 @@
+export const enable_status_indicators: FeatureToggle = {
+  name: "Show/Hide Status Indicators",
+  category: "GAMEPLAY",
+  description: "This shows/hides status indicators.",
```
```suggestion
  description: "This toggles whether or not you will see status indicators.",
```

GoldenAlpharex commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,198 @@
+/datum/preference/toggle/enable_status_indicators/create_default_value()
+	return TRUE
+
+/datum/preference/toggle/enable_status_indicators/apply_to_client(client/client, value)
+	. = ..()
+	var/atom/movable/screen/plane_master/runechat/status/status = locate() in client?.screen
+	if(!status)
+		return
+	if(value)
+		status.alpha = 255
+	else
+		status.alpha = 0
```
I'm not entirely sure why, but I've got a sneaky suspicion this isn't exactly working, considering I've had my pref off since it's been added, and I still see them now.

nevimer replied:
Think I should delete this layer from mobs instead, and recreate?

GoldenAlpharex reviewed (changes requested):
There's a bit more work that needs to be done in here, so we avoid running into issues of performance in the long run.

Furthermore, I think there should be a Lag switch for this, that would work exactly like how the runechat lagswitch works, because I'm *sure* there's a cost attached to this, and that there might be situations where we'd rather not have them to save up on some performance.

nevimer commented:
> reviews

doing them on a seperate branch that im going to squash merge into this one so I break this working state

nevimer commented:
Doing reviews tomorrow either before or after another PR's reviews

nevimer commented on modular_skyrat/master_files/code/modules/pain_shock/overrides.dm:
```diff
@@ -0,0 +1,78 @@
+/datum/status_effect/incapacitating/stun/on_apply()
+	. = ..()
+	owner.add_status_indicator("stunned") // SKYRAT EDIT ADDITION: Status indicators
+
+/datum/status_effect/incapacitating/stun/on_remove()
+	. = ..()
+	owner.remove_status_indicator("stunned") // SKYRAT EDIT ADDITION: Status indicators
+
+
+/datum/status_effect/incapacitating/knockdown/on_apply()
+	. = ..()
+	owner.add_status_indicator("weakened") // SKYRAT EDIT ADDITION: Status indicators
+
+/datum/status_effect/incapacitating/knockdown/on_remove()
+	. = ..()
+	if(id != STAT_TRAIT)
+		owner.remove_status_indicator("weakened") // SKYRAT EDIT ADDITION: Status indicators
+
+/datum/status_effect/incapacitating/immobilized/on_apply()
+	. = ..()
+	owner.add_status_indicator("weakened")
+/datum/status_effect/incapacitating/immobilized/on_remove()
+	. = ..()
+	if(id != STAT_TRAIT)
+		owner.remove_status_indicator("weakened") // SKYRAT EDIT ADDITION: Status indicators
+
+/datum/status_effect/incapacitating/paralyzed/on_apply()
+	. = ..()
+	owner.add_status_indicator("paralysis") // SKYRAT EDIT ADDITION: Status indicators
+
+/datum/status_effect/incapacitating/paralyzed/on_remove()
+	. = ..()
+	owner.remove_status_indicator("paralysis") // SKYRAT EDIT ADDITION: Status indicators
```
Early returns if the indicator exists

nevimer commented on modular_skyrat/master_files/code/modules/pain_shock/overrides.dm:
```diff
@@ -0,0 +1,78 @@
+
+/* */
+
+/datum/status_effect/incapacitating/incapacitated/on_apply()
+	. = ..()
+	owner.add_status_indicator("weakened") // SKYRAT EDIT ADDITION: Status indicators
+
+
+
+/datum/status_effect/incapacitating/incapacitated/on_remove()
+	. = ..()
+	owner.remove_status_indicator("weakened") // SKYRAT EDIT ADDITION: Status indicators
+
+
+/datum/status_effect/incapacitating/unconscious/on_apply()
+	. = ..()
+	owner.add_status_indicator("sleeping") // SKYRAT EDIT ADDITION: Status indicators
+
+/datum/status_effect/incapacitating/unconscious/on_remove()
+	. = ..()
+	owner.remove_status_indicator("sleeping") // SKYRAT EDIT ADDITION: Status indicators
```
gone

nevimer commented on modular_skyrat/master_files/code/modules/pain_shock/overrides.dm:
```diff
@@ -0,0 +1,78 @@
+	. = ..()
+	owner.remove_status_indicator("paralysis") // SKYRAT EDIT ADDITION: Status indicators
+
+
+/* */
```
gone in next commit

nevimer commented on modular_skyrat/master_files/code/modules/pain_shock/overrides.dm:
```diff
@@ -0,0 +1,78 @@
+/datum/status_effect/incapacitating/stun/on_apply()
+	. = ..()
+	owner.add_status_indicator("stunned") // SKYRAT EDIT ADDITION: Status indicators
```
defines in next commit

nevimer commented:
Ready for re-review. Might need help with the bigger asks.

YakumoChen commented:
I'm seeing problems where statuses will just stick around and never go away after you get them

nevimer commented:
> I'm seeing problems where statuses will just stick around and never go away after you get them

I'm looking for ways to reproduce it. So far, it seems to be a race condition... And my ways to fix it don't seem to be working. Hm.

nevimer commented:
Alright, I have NEW IDEAS on how to do the logic on this, so, I will be doing a big update to this PR with enough coffee provided.

Weblure commented:
The issues I see with this are:

- This hurts stealth in two ways. First, you're broadcasting information to people that would've otherwise been subtle or required an examination. The second way is much more litteral: these icons stand out like the sun when you're trying be sneaky. It makes it very difficult to try to blend into the shadows or hide under a person/sofa when you have neon lights over your head. With all the stealth nerfs we've been getting lately, even something as small as this can be a major blow.
- The preference toggle isn't very useful if it only turns it off for yourself locally and other people can still see your statuses above your head. In fact, by turning it off, you are only putting yourself at a disadvantage. I'm not sure exactly what this toggle actually does, since there's nothing written about it here or in the tooltip.

nevimer commented:
> The issues I see with this are:
> 
>     * This hurts stealth in two ways. First, you're broadcasting information to people that would've otherwise been subtle or required an examination. The second way is much more litteral: these icons stand out like the sun when you're trying be sneaky. It makes it very difficult to try to blend into the shadows or hide under a person/sofa when you have neon lights over your head. With all the stealth nerfs we've been getting lately, even something as small as this can be a major blow.
> 
>     * The preference toggle would be more useful if it didn't only turn it off for yourself locally, i.e. if it'd actually prevent other people from seeing your statuses above your head. In fact, by turning it off, you'd only be putting yourself at a disadvantage. I'm not sure exactly what this toggle actually does, since there's nothing written about it here or in the tooltip.

Well, CI, and other things don't work with stealth things like ninja cloaking either. And the FOV thing I am not too woried about because.. 

1. TG has it in their crosshairs for removal, last I heard. They don't even run it afaik.
2. We have hearing indicators when you're blind, indicating the position of noises and the action. 
3. Turning it off for yourself is a disadvantage, yes, but so is turning off widescreen mode, fake AO, runechat, and every other feature added to the game that makes it more palatable. 

Working with the FOV code looks extremely painful, and for the sake of modularity and its potential impending removal I want to avoid writing a dependency on it. 

I'll probably make it so that indicators disappear when you have any form of alpha going on, or other edge cases since I've got a list now.

YakumoChen commented:
Easy fix to just put the indicators on the same plane as character sprites or whatever it takes to give them the same alpha and visibility as, say, HUDs. Base HUD icons share visibility and vanish when in darkness or lowered alpha.

Panicode commented:
No idea why people are being so upset about this?
If someone is stunned IRL, they'd be obviously stunned-looking.
We're staring at a pixel person. We have no idea how they're feeling at a glance; If we're able to examine them actively to see how they're feeling: then we should conversely be able to graphically see them under that status effect.

I don't how how this would impact being "stealthy" as, if you were being under most of these statuses, you were already caught, or under some sort of stimulants.
If people really wanted this to be a specific mechanic, then maybe having 'walk mode' on could force-hide it?

jjpark-kb commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,187 @@
+/mob/living/carbon/Initialize(mapload)
+	. = ..()
+	// The Basics
+	RegisterSignal(src, COMSIG_CARBON_HEALTH_UPDATE, .proc/status_indicator_evaluate)
+	RegisterSignal(src, COMSIG_LIVING_DEATH, .proc/cut_indicators_overlays)
+	// Catching weird cases
+	RegisterSignal(src, COMSIG_CARBON_HELP_ACT, .proc/status_indicator_evaluate)
+	RegisterSignal(src, COMSIG_LIVING_SET_BODY_POSITION, .proc/status_indicator_evaluate)
+	// When things actually happen
+	RegisterSignal(src, COMSIG_LIVING_STATUS_STUN, .proc/status_indicator_evaluate)
+	RegisterSignal(src, COMSIG_LIVING_STATUS_KNOCKDOWN, .proc/status_indicator_evaluate)
+	RegisterSignal(src, COMSIG_LIVING_STATUS_PARALYZE, .proc/status_indicator_evaluate)
+	RegisterSignal(src, COMSIG_LIVING_STATUS_IMMOBILIZE, .proc/status_indicator_evaluate)
+	RegisterSignal(src, COMSIG_LIVING_STATUS_UNCONSCIOUS, .proc/status_indicator_evaluate)
+	RegisterSignal(src, COMSIG_LIVING_STATUS_SLEEP, .proc/status_indicator_evaluate)
```
pretty sure there is a way to cut down on this.

nevimer replied:
Ah, man, these are all the signals used where I would want to evaluate the mob. I think it's better than over-riding procs.

nevimer replied:
I've cut a few! let me know!

jjpark-kb commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,187 @@
+/// Finds a status indicator on a mob.
+/mob/living/proc/get_status_indicator(image/prospective_indicator)
+
+	for(var/image/I in status_indicators)
+		if(I.icon_state == prospective_indicator)
+			return I
```
single letter vars.

nevimer replied:
done!

jjpark-kb commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,187 @@
+
+	// Now the indicator row can actually be built.
+	for(var/prospective_indicator in status_indicators)
+		var/image/I = prospective_indicator
+
+		// This is a semi-HUD element, in a similar manner as medHUDs, in that they're 'above' everything else in the world,
+		// but don't pierce obfuscation layers such as blindness or darkness, unlike actual HUD elements like inventory slots.
+		I.plane = PLANE_STATUS_INDICATOR
+		I.layer = STATUS_LAYER
+		I.appearance_flags = PIXEL_SCALE|TILE_BOUND|NO_CLIENT_COLOR|RESET_COLOR|RESET_ALPHA|RESET_TRANSFORM|KEEP_APART
+		I.pixel_y = y_offset
+		I.pixel_x = current_x_position
+		add_overlay(I)
```
single letter vars.

nevimer replied:
fixed!

jjpark-kb commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,187 @@
+
+/mob/living/proc/get_icon_scale(livingmob)
+	if(!iscarbon(livingmob)) // normal mobs are always 1 for scale - hopefully all borgs and simplemobs get this one
+		return 1
```
make this a define like #define DEFAULT_MOB_SCALE 1

nevimer replied:
done!

jjpark-kb commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,187 @@
+	if(!iscarbon(livingmob)) // normal mobs are always 1 for scale - hopefully all borgs and simplemobs get this one
+		return 1
+	var/mob/living/carbon/passed_mob = livingmob // we're possibly a player! We have size prefs!
+	var/mysize = (passed_mob.dna?.current_body_size ? passed_mob.dna.current_body_size : 1)
```
1 again would be the define

nevimer replied:
done!

jjpark-kb commented on modular_skyrat/master_files/code/modules/pain_shock/status_indicator.dm:
```diff
@@ -0,0 +1,187 @@
+
+	if(stat == DEAD)
+		return
+	var/mob/living/carbon/carbon = src
```
more descriptive var name, such as carbon_src

nevimer replied:
done!
